### PR TITLE
[B-03960] Peer Query

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/holochain/pull/827

- Query for peers covering location
- Query for a specific peer
- discovery logic for rpc_single to loop over discovering a specific peer

### CODE REVIEW

File: crates/holochain/src/local_network_tests.rs
Comment (maackle) [inline crates/holochain/src/local_network_tests.rs:37]: OK but let's see if it works on CI

```suggestion
```

File: crates/holochain/tests/agent_scaling.rs
Comment (maackle) [inline crates/holochain/tests/agent_scaling.rs:86]: same

```suggestion
```

**Review by maackle — COMMENTED** [review]
Still reviewing, but want to run these ignored tests in CI

File: crates/kitsune_p2p/types/src/timeout.rs
Comment (maackle) [inline crates/kitsune_p2p/types/src/timeout.rs:138]: I'm surprised that `times` would be over 4. `2 + 4 + 8 + 16 + 32` is way over 15. Am I reading this wrong? Or are you *expecting* race conditions in the backoff for this test to work?

Comment (neonphog) [reply crates/kitsune_p2p/types/src/timeout.rs:138]: so, yeah, the initial is `2` ms, the max is `15` ms, and we're letting the loop run for `100` ms, and seeing how many times are pushed on to the array... so, assuming that `.await` returns at exactly the backoff timeout without any additional delay, we'd get:

```rust
[
  0, // initial
  2, // after 2 ms backoff
  6, // after 4 ms backoff
  14, // after 8 ms backoff
  29, // after 15 ms backoff
  44, // after 15 ms backoff
  59, // after 15 ms backoff
  74, // after 15 ms backoff
  89, // after 15 ms backoff
].len() == 9 // entries
```

File: crates/kitsune_p2p/direct_test/src/lib.rs
Comment (neonphog) [reply crates/kitsune_p2p/direct_test/src/lib.rs]: haha, totally : )

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs]: done

File: crates/kitsune_p2p/direct/src/persist_mem.rs
Comment (maackle) [inline crates/kitsune_p2p/direct/src/persist_mem.rs:375]: why not do this right now?

File: crates/holochain_sqlite/src/sql/p2p_agent_store/query_near_basis.sql
Comment (maackle) [inline crates/holochain_sqlite/src/sql/p2p_agent_store/query_near_basis.sql:9]: I wish my PR that simplified arc queries down to a single range had already been merged :_(

**Review by maackle — COMMENTED** [review]
A few more questions about discarded errors, and a logic question

File: crates/holochain_sqlite/src/sql/p2p_agent_store/query_near_basis.sql
Comment (maackle) [reply crates/holochain_sqlite/src/sql/p2p_agent_store/query_near_basis.sql:9]: I guess I will wait to merge this until after this PR is merged https://github.com/holochain/holochain/pull/816

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:164]: yep, good catch, applying.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (freesig) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:74]: There's a good chance this query will be slow due to p2p_store write contention. Is it ok for the back pressure to comeback onto this function?

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (freesig) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:184]: This rng is happening inside a double nested loop. Might be worth reusing it?

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (freesig) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:228]: So this is happening in the context of a single_rpc. Is an rpc able to wait long enough to handle exponentially backing off?

File: crates/kitsune_p2p/kitsune_p2p/src/test_util/harness_agent.rs
Comment (freesig) [inline crates/kitsune_p2p/kitsune_p2p/src/test_util/harness_agent.rs:169]: Should this be done now? It's pretty easy to sort a vec and I don't think we'll ever have enough peers for sorting the vec to be too slow.

**Review by freesig — APPROVED** [review]
Looking good overall. A few minor suggestions. Trying to keep in mind parts of the code that are likely to take a while to complete and make sure that it's appropriate for the calling functions to really wait that long.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs]: done

File: crates/kitsune_p2p/kitsune_p2p/src/test_util/harness_agent.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/test_util/harness_agent.rs:169]: This is the deprecated test harness, I don't want to spend any time maintaining it...

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:190]: No, we cannot cache thread-local data across awaits, they are not Send, because they are thread local... also thread-local is very fast, you don't need any locks around it, because there cannot be contention : )

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:74]: Yeah, `search_and_discover_peer_connect` happens in the task context of whatever conductor call is making the request... so long as conductor is making parallel calls, they will happen in parallel.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:234]: The backoff itself doesn't have a problem... the single_rpc timeout will be something like 30 seconds, and this backoff starts at 100ms up to a max of 1000ms... so it should get quite a few tries in...

EXCEPT... each internal call to things like `peer_connect` will use the full remaining timeout, so if the first returned node data we get happens to be to a node that is offline, we'll end up timing out without trying any additional nodes.

Again, I'm not sure a better way to solve this without making requests just dangle indefinitely if we've got that QUIC idle timeout set to 30 seconds...

File: crates/kitsune_p2p/direct/src/persist_mem.rs
Comment (neonphog) [reply crates/kitsune_p2p/direct/src/persist_mem.rs]: done

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs
Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs:274]: I would love to see every result either captured, or have a comment explaining why we don't need to capture it, so we don't have to think about it anymore. We've had cases come up where we did discard a potentially useful error message, so now I'm always on lookout for that. So what do you think about just adding comments here?

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs:234]: We're going to call this okay for now, we may need to address this differently in the future.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs]: Ok, added the comment as a macro, so I don't have to say the same thing 8 times : )

**Pull request merged**
